crates/telio-starcast/src/nat.rs

```diff
@@ -0,0 +1,648 @@
+    PacketTooShort,
+    /// The packet has an unexpected IP version
+    UnexpectedIpVersion,
+    /// The packet has an unexpected transport protocol
+    UnexpectedTransportProtocol,
+    /// There is no mapping for the given port
+    MappingNotFound,
+}
+
+pub trait Nat {
+    /// Translate outgoing packet
+    /// Change the destination to the peer's original ip and port
+    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+    /// Translate incomming packet
+    /// Change the source to the Starcast's virtual ip and natted port
+    fn translate_incomming(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+}
+
+type Mapper = fn(src_addr: SocketAddr) -> u16;
+
+fn default_mapper(src_addr: SocketAddr) -> u16 {
+    let mut hasher = DefaultHasher::new();
+    match src_addr.ip() {
+        IpAddr::V4(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+        IpAddr::V6(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+    };
+    src_addr.port().hash(&mut hasher);
+    hasher.finish() as u16 // TODO: think if we want to use the full 16 bits or maybe some range e.g. >=1024
+}
+
+pub struct StarcastNat {
+    mappings: LruCache<u16, SocketAddr>,
+    vpeer_ipv4_addr: Ipv4Addr,
+    vpeer_ipv6_addr: Ipv6Addr,
+    mapper: Mapper,
+}
+
+impl StarcastNat {
+    /// Create new starcast nat with default capcity of 2048 entries and 2 minute timeout
+    /// 2min default is taken from (https://datatracker.ietf.org/doc/html/rfc4787#section-4.3)
+    pub fn new(vpeer_ipv4: Ipv4Addr, vpeer_ipv6: Ipv6Addr) -> Self {
+        const LRU_CAPACITY: usize = 2048;
+        const LRU_TIMEOUT: Duration = Duration::from_secs(120);
+        Self::new_custom(
+            vpeer_ipv4,
+            vpeer_ipv6,
+            LRU_CAPACITY,
+            LRU_TIMEOUT,
+            default_mapper,
+        )
+    }
+
+    /// Create new starcast nat with custom capacity and timeout
+    fn new_custom(
+        vpeer_ipv4_addr: Ipv4Addr,
+        vpeer_ipv6_addr: Ipv6Addr,
+        capacity: usize,
+        timeout: Duration,
+        mapper: Mapper,
+    ) -> Self {
+        Self {
+            mappings: LruCache::new(timeout, capacity),
+            vpeer_ipv4_addr,
+            vpeer_ipv6_addr,
+            mapper,
+        }
+    }
+
+    fn translate_incomming_internal<'a, P: MutableIpPacket<'a>>(
+        &mut self,
+        packet: &'a mut [u8],
+    ) -> Result<(), Error> {
+        let mut ip_packet = P::new(packet).ok_or(Error::PacketTooShort)?;
+        if ip_packet.get_next_level_protocol() != IpNextHeaderProtocols::Udp {
+            return Err(Error::UnexpectedTransportProtocol);
+        }
+
+        let src_ip = ip_packet.get_source();
+        let dst_ip = ip_packet.get_destination();
+
+        match src_ip {
+            IpAddr::V4(_) => {
+                ip_packet.set_source(IpAddr::V4(self.vpeer_ipv4_addr));
+                ip_packet.set_checksum(ipv4::checksum(&ip_packet.to_immutable_ipv4()));
```

Comment:
So now we do a full recalculation of the checksum, it's the easiest way with api provided, but not the most unoptimal.
It's smarter to just ajust checksum, see https://datatracker.ietf.org/doc/html/rfc1141 and https://datatracker.ietf.org/doc/html/rfc1624

Reply:
I would propose to create an improvement task for this purpose and leave it as is for now

Reply:
I would be more for adding it now, it should not be that hard, expecialy if you have unit tests for it

Reply:
Added. Please take a look